File: labs/aula5/PEMKeyGenerator.py
import sys
from cryptography.hazmat.primitives.asymmetric import rsa
from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.backends import default_backend

def main():
    if len(sys.argv) < 3:
        print ( "Usage: PEMKeyGenerator bits password [file]\n" )
        sys.exit( 1 )

    key_size = int( sys.argv[1] )
    pwd = sys.argv[2]

    if len(sys.argv) == 4:
        key_file = open( sys.argv[3], "wb" )
    else:
        key_file = sys.stdout.buffer;

    # Use 65537 (2^16 + 1) as public exponent

    keyPair = rsa.generate_private_key( 65537, key_size, default_backend() )

    # Save key pair to a PEM file protected by a password
    # (PKCS #8 password-based, private key encryption)

    pemEncoding = keyPair.private_bytes( serialization.Encoding.PEM,
                                            serialization.PrivateFormat.PKCS8,
                                            serialization.BestAvailableEncryption( bytes( pwd, "utf-8" ) ) )

    key_file.write( pemEncoding )

if __name__ == "__main__":
    main()

File: labs/aula5/PEMKeyGenerator.java
import java.security.*;
import java.io.*;

import org.bouncycastle.jce.provider.BouncyCastleProvider;
import org.bouncycastle.util.io.pem.*;
import org.bouncycastle.operator.*;
import org.bouncycastle.openssl.*;
import org.bouncycastle.openssl.jcajce.*;

import static java.lang.System.*;

class PEMKeyGenerator {

static public void 
main ( String[] args )
{
    OutputStream ostream = null;

    // Chck parameters

    if (args.length < 2) {
        err.println ( "Usage: PEMKeyGenerator bits password [file]" );
        System.exit( 1 );
    }

    int keySize = Integer.parseInt( args[0] );
    String pwd = args[1];

    try {

    if (args.length == 3) {
        ostream = new FileOutputStream( args[2] );
    }
    else {
        ostream = out;
    }

    // Generate key pair

    KeyPairGenerator kpg = KeyPairGenerator.getInstance( "RSA" );

    kpg.initialize( keySize );
    KeyPair kp = kpg.generateKeyPair();

    // AddtheBouncy Castle provider

    Security.addProvider( new BouncyCastleProvider() );

    // construct encryptor to encrypt the private key

    JceOpenSSLPKCS8EncryptorBuilder encBuilder =
        new JceOpenSSLPKCS8EncryptorBuilder( PKCS8Generator.AES_128_CBC );  

    encBuilder.setRandom( new SecureRandom() );  
    encBuilder.setPasssword( pwd.toCharArray() );
    OutputEncryptor encryptor = encBuilder.build();

    // Construct generator to create the PKCS8 object from the private key and encryptor

    JcaPKCS8Generator pkcsGenerator = new JcaPKCS8Generator( kp.getPrivate(), encryptor );  
    PemObject encryptedEncodedKey = pkcsGenerator.generate();  

    // Write the PKCS #8 object to a PEM file

    JcaPEMWriter keyFile = new JcaPEMWriter( new OutputStreamWriter ( ostream ) );
    keyFile.writeObject( encryptedEncodedKey );
    keyFile.close ();

    } catch(Exception e) {
        err.println( "Exception: " + e );
    }
}

}
